// ramda-extension/es/index.js
/* generated by plop */
// aliases for backward compatibility
import _equalsZero from './equalsToZero';
export { _equalsZero as equalsZero };
import _equalsEmptyString from './equalsToEmptyString';
export { _equalsEmptyString as equalsEmptyString };
import _notNil from './notNil';
export { _notNil as notNil };
import _notEmpty from './notEmpty';
export { _notEmpty as notEmpty };
import _notNaN from './notNaN';
export { _notNaN as notNaN };
import _notNumeric from './notNumeric';
export { _notNumeric as notNumeric };
import _alwaysNull from './alwaysNull';
export { _alwaysNull as alwaysNull };
import _alwaysEmptyString from './alwaysEmptyString';
export { _alwaysEmptyString as alwaysEmptyString };
import _alwaysEmptyArray from './alwaysEmptyArray';
export { _alwaysEmptyArray as alwaysEmptyArray };
import _alwaysZero from './alwaysZero';
export { _alwaysZero as alwaysZero };
import _alwaysOne from './alwaysOne';
export { _alwaysOne as alwaysOne };
import _alwaysEmptyObject from './alwaysEmptyObject';
export { _alwaysEmptyObject as alwaysEmptyObject };
import _applyCompose from './applyCompose';
export { _applyCompose as applyCompose };
import _applyPipe from './applyPipe';
export { _applyPipe as applyPipe };
import _between from './between';
export { _between as between };
import _isArray from './isArray';
export { _isArray as isArray };
import _isFunction from './isFunction';
export { _isFunction as isFunction };
import _isObject from './isObject';
export { _isObject as isObject };
import _isNegative from './isNegative';
export { _isNegative as isNegative };
import _isNilOrEmptyString from './isNilOrEmptyString';
export { _isNilOrEmptyString as isNilOrEmptyString };
import _isNotEmpty from './isNotEmpty';
export { _isNotEmpty as isNotEmpty };
import _isNotNaN from './isNotNaN';
export { _isNotNaN as isNotNaN };
import _isNotNil from './isNotNil';
export { _isNotNil as isNotNil };
import _isNotNilObject from './isNotNilObject';
export { _isNotNilObject as isNotNilObject };
import _isNotNumeric from './isNotNumeric';
export { _isNotNumeric as isNotNumeric };
import _isNotRegExp from './isNotRegExp';
export { _isNotRegExp as isNotRegExp };
import _isNumeric from './isNumeric';
export { _isNumeric as isNumeric };
import _isPositive from './isPositive';
export { _isPositive as isPositive };
import _isPromise from './isPromise';
export { _isPromise as isPromise };
import _isString from './isString';
export { _isString as isString };
import _isRegExp from './isRegExp';
export { _isRegExp as isRegExp };
import _isTruthy from './isTruthy';
export { _isTruthy as isTruthy };
import _isFalsy from './isFalsy';
export { _isFalsy as isFalsy };
import _isError from './isError';
export { _isError as isError };
import _log from './log';
export { _log as log };
import _trace from './trace';
export { _trace as trace };
import _reduceCallable from './reduceCallable';
export { _reduceCallable as reduceCallable };
import _reduceSource from './reduceSource';
export { _reduceSource as reduceSource };
import _findNotNil from './findNotNil';
export { _findNotNil as findNotNil };
import _dispatch from './dispatch';
export { _dispatch as dispatch };
import _dispatchWith from './dispatchWith';
export { _dispatchWith as dispatchWith };
import _noop from './noop';
export { _noop as noop };
import _startsWithPrefix from './startsWithPrefix';
export { _startsWithPrefix as startsWithPrefix };
import _startsWithPrefixIgnoreCase from './startsWithPrefixIgnoreCase';
export { _startsWithPrefixIgnoreCase as startsWithPrefixIgnoreCase };
import _endsWithSuffix from './endsWithSuffix';
export { _endsWithSuffix as endsWithSuffix };
import _endsWithSuffixIgnoreCase from './endsWithSuffixIgnoreCase';
export { _endsWithSuffixIgnoreCase as endsWithSuffixIgnoreCase };
import _argumentsToList from './argumentsToList';
export { _argumentsToList as argumentsToList };
import _containsAll from './containsAll';
export { _containsAll as containsAll };
import _containsAny from './containsAny';
export { _containsAny as containsAny };
import _containsNone from './containsNone';
export { _containsNone as containsNone };
import _unfoldObjectDots from './unfoldObjectDots';
export { _unfoldObjectDots as unfoldObjectDots };
import _toUpperFirst from './toUpperFirst';
export { _toUpperFirst as toUpperFirst };
import _toLowerFirst from './toLowerFirst';
export { _toLowerFirst as toLowerFirst };
import _toCamelCase from './toCamelCase';
export { _toCamelCase as toCamelCase };
import _toPascalCase from './toPascalCase';
export { _toPascalCase as toPascalCase };
import _toSnakeCase from './toSnakeCase';
export { _toSnakeCase as toSnakeCase };
import _toKebabCase from './toKebabCase';
export { _toKebabCase as toKebabCase };
import _toDotCase from './toDotCase';
export { _toDotCase as toDotCase };
import _toScreamingSnakeCase from './toScreamingSnakeCase';
export { _toScreamingSnakeCase as toScreamingSnakeCase };
import _listToString from './listToString';
export { _listToString as listToString };
import _splitByNonAlphaNumeric from './splitByNonAlphaNumeric';
export { _splitByNonAlphaNumeric as splitByNonAlphaNumeric };
import _constructRegExp from './constructRegExp';
export { _constructRegExp as constructRegExp };
import _rejectNil from './rejectNil';
export { _rejectNil as rejectNil };
import _rejectEq from './rejectEq';
export { _rejectEq as rejectEq };
import _splitByDot from './splitByDot';
export { _splitByDot as splitByDot };
import _joinWithDot from './joinWithDot';
export { _joinWithDot as joinWithDot };
import _joinWithSpace from './joinWithSpace';
export { _joinWithSpace as joinWithSpace };
import _joinWithUnderscore from './joinWithUnderscore';
export { _joinWithUnderscore as joinWithUnderscore };
import _joinWithDash from './joinWithDash';
export { _joinWithDash as joinWithDash };
import _dissocDotPath from './dissocDotPath';
export { _dissocDotPath as dissocDotPath };
import _overHead from './overHead';
export { _overHead as overHead };
import _dotPath from './dotPath';
export { _dotPath as dotPath };
import _assocDotPath from './assocDotPath';
export { _assocDotPath as assocDotPath };
import _mergeWithDotPath from './mergeWithDotPath';
export { _mergeWithDotPath as mergeWithDotPath };
import _mapKeys from './mapKeys';
export { _mapKeys as mapKeys };
import _applyIfNotNil from './applyIfNotNil';
export { _applyIfNotNil as applyIfNotNil };
import _equalsLength from './equalsLength';
export { _equalsLength as equalsLength };
import _equalsStringIgnoreCase from './equalsStringIgnoreCase';
export { _equalsStringIgnoreCase as equalsStringIgnoreCase };
import _gteThanLength from './gteThanLength';
export { _gteThanLength as gteThanLength };
import _gtThanLength from './gtThanLength';
export { _gtThanLength as gtThanLength };
import _lteThanLength from './lteThanLength';
export { _lteThanLength as lteThanLength };
import _ltThanLength from './ltThanLength';
export { _ltThanLength as ltThanLength };
import _viewEq from './viewEq';
export { _viewEq as viewEq };
import _viewWith from './viewWith';
export { _viewWith as viewWith };
import _flattenArgs from './flattenArgs';
export { _flattenArgs as flattenArgs };
import _cx from './cx';
export { _cx as cx };
import _propNotEq from './propNotEq';
export { _propNotEq as propNotEq };
import _pathNotEq from './pathNotEq';
export { _pathNotEq as pathNotEq };
import _duplicate from './duplicate';
export { _duplicate as duplicate };
import _replicate from './replicate';
export { _replicate as replicate };
import _keyMirror from './keyMirror';
export { _keyMirror as keyMirror };
import _valueMirror from './valueMirror';
export { _valueMirror as valueMirror };
import _headArg from './headArg';
export { _headArg as headArg };
import _lastArg from './lastArg';
export { _lastArg as lastArg };
import _mapKeysWithValue from './mapKeysWithValue';
export { _mapKeysWithValue as mapKeysWithValue };
import _camelizeKeys from './camelizeKeys';
export { _camelizeKeys as camelizeKeys };
import _includes from './includes';
export { _includes as includes };
import _notInclude from './notInclude';
export { _notInclude as notInclude };
import _pipeC from './pipeC';
export { _pipeC as pipeC };
import _composeC from './composeC';
export { _composeC as composeC };
import _padLeftUntil from './padLeftUntil';
export { _padLeftUntil as padLeftUntil };
import _padLeft from './padLeft';
export { _padLeft as padLeft };
import _padRightUntil from './padRightUntil';
export { _padRightUntil as padRightUntil };
import _padRight from './padRight';
export { _padRight as padRight };
import _xPairs from './xPairs';
export { _xPairs as xPairs };
import _xPairsRight from './xPairsRight';
export { _xPairsRight as xPairsRight };
import _memoizeWithIdentity from './memoizeWithIdentity';
export { _memoizeWithIdentity as memoizeWithIdentity };
import _objOfOver from './objOfOver';
export { _objOfOver as objOfOver };
import _mapKeysAndValues from './mapKeysAndValues';
export { _mapKeysAndValues as mapKeysAndValues };
import _mergeDeepLeftAll from './mergeDeepLeftAll';
export { _mergeDeepLeftAll as mergeDeepLeftAll };
import _mergeDeepRightAll from './mergeDeepRightAll';
export { _mergeDeepRightAll as mergeDeepRightAll };
import _mergeDeepAllWith from './mergeDeepAllWith';
export { _mergeDeepAllWith as mergeDeepAllWith };
import _mergeDeepAllWithKey from './mergeDeepAllWithKey';
export { _mergeDeepAllWithKey as mergeDeepAllWithKey };
import _defaultToEmptyArray from './defaultToEmptyArray';
export { _defaultToEmptyArray as defaultToEmptyArray };
import _defaultToEmptyObject from './defaultToEmptyObject';
export { _defaultToEmptyObject as defaultToEmptyObject };
import _defaultToEmptyString from './defaultToEmptyString';
export { _defaultToEmptyString as defaultToEmptyString };
import _defaultToFalse from './defaultToFalse';
export { _defaultToFalse as defaultToFalse };
import _defaultToOne from './defaultToOne';
export { _defaultToOne as defaultToOne };
import _defaultToTrue from './defaultToTrue';
export { _defaultToTrue as defaultToTrue };
import _defaultToZero from './defaultToZero';
export { _defaultToZero as defaultToZero };
import _equalsToEmptyArray from './equalsToEmptyArray';
export { _equalsToEmptyArray as equalsToEmptyArray };
import _equalsToEmptyObject from './equalsToEmptyObject';
export { _equalsToEmptyObject as equalsToEmptyObject };
import _equalsToEmptyString from './equalsToEmptyString';
export { _equalsToEmptyString as equalsToEmptyString };
import _equalsToFalse from './equalsToFalse';
export { _equalsToFalse as equalsToFalse };
import _equalsToOne from './equalsToOne';
export { _equalsToOne as equalsToOne };
import _equalsToNull from './equalsToNull';
export { _equalsToNull as equalsToNull };
import _equalsToTrue from './equalsToTrue';
export { _equalsToTrue as equalsToTrue };
import _equalsToZero from './equalsToZero';
export { _equalsToZero as equalsToZero };
import _notEqual from './notEqual';
export { _notEqual as notEqual };
import _notEqualToEmptyArray from './notEqualToEmptyArray';
export { _notEqualToEmptyArray as notEqualToEmptyArray };
import _notEqualToEmptyObject from './notEqualToEmptyObject';
export { _notEqualToEmptyObject as notEqualToEmptyObject };
import _notEqualToEmptyString from './notEqualToEmptyString';
export { _notEqualToEmptyString as notEqualToEmptyString };
import _notEqualToFalse from './notEqualToFalse';
export { _notEqualToFalse as notEqualToFalse };
import _notEqualToNull from './notEqualToNull';
export { _notEqualToNull as notEqualToNull };
import _notEqualToOne from './notEqualToOne';
export { _notEqualToOne as notEqualToOne };
import _notEqualToTrue from './notEqualToTrue';
export { _notEqualToTrue as notEqualToTrue };
import _notEqualToZero from './notEqualToZero';
export { _notEqualToZero as notEqualToZero };
import _toggle from './toggle';
export { _toggle as toggle };
import _dotPathEq from './dotPathEq';
export { _dotPathEq as dotPathEq };
import _dotPathNotEq from './dotPathNotEq';
export { _dotPathNotEq as dotPathNotEq };
import _dotPathOr from './dotPathOr';
export { _dotPathOr as dotPathOr };
import _dotPathSatisfies from './dotPathSatisfies';
export { _dotPathSatisfies as dotPathSatisfies };
import _dotPathNotSatisfies from './dotPathNotSatisfies';
export { _dotPathNotSatisfies as dotPathNotSatisfies };
import _propNotSatisfies from './propNotSatisfies';
export { _propNotSatisfies as propNotSatisfies };
import _pathNotSatisfies from './pathNotSatisfies';
export { _pathNotSatisfies as pathNotSatisfies };
import _mapIndexed from './mapIndexed';
export { _mapIndexed as mapIndexed };
import _isNilOrEmpty from './isNilOrEmpty';
export { _isNilOrEmpty as isNilOrEmpty };
import _isNumber from './isNumber';
export { _isNumber as isNumber };
import _isPlainObject from './isPlainObject';
export { _isPlainObject as isPlainObject };
import _applyIfFunction from './applyIfFunction';
export { _applyIfFunction as applyIfFunction };
import _callIfFunction from './callIfFunction';
export { _callIfFunction as callIfFunction };
import _flipIncludes from './flipIncludes';
export { _flipIncludes as flipIncludes };
import _notFlipInclude from './notFlipInclude';
export { _notFlipInclude as notFlipInclude };
import _isBoolean from './isBoolean';
export { _isBoolean as isBoolean };
import _ensureArray from './ensureArray';
export { _ensureArray as ensureArray };
import _toEntries from './toEntries';
export { _toEntries as toEntries };
import _get from './get';
export { _get as get };
import _uniqMap from './uniqMap';
export { _uniqMap as uniqMap };
import _uniqFilter from './uniqFilter';
export { _uniqFilter as uniqFilter };
import _uniqLength from './uniqLength';
export { _uniqLength as uniqLength };
import _uniqTake from './uniqTake';
export { _uniqTake as uniqTake };
import _uniqTakeLast from './uniqTakeLast';
export { _uniqTakeLast as uniqTakeLast };
import _uniqTakeWhile from './uniqTakeWhile';
export { _uniqTakeWhile as uniqTakeWhile };
import _uniqDrop from './uniqDrop';
export { _uniqDrop as uniqDrop };
import _uniqTakeLastWhile from './uniqTakeLastWhile';
export { _uniqTakeLastWhile as uniqTakeLastWhile };
import _uniqDropLast from './uniqDropLast';
export { _uniqDropLast as uniqDropLast };
import _uniqDropWhile from './uniqDropWhile';
export { _uniqDropWhile as uniqDropWhile };
import _uniqDropLastWhile from './uniqDropLastWhile';
export { _uniqDropLastWhile as uniqDropLastWhile };
import _uniqAppend from './uniqAppend';
export { _uniqAppend as uniqAppend };
import _uniqPrepend from './uniqPrepend';
export { _uniqPrepend as uniqPrepend };
import _groupByPath from './groupByPath';
export { _groupByPath as groupByPath };
import _groupByProp from './groupByProp';
export { _groupByProp as groupByProp };
